File: typescipt/src/parser/iVariableContext.ts


export interface IVariableContext {
   void AddVariable(string variableName, VariableType type, VariableSource source);

   void RegisterVariableAndVerifyUnique(SourceReference reference, string variableName, VariableType type,
     VariableSource source);

   boolean EnsureVariableExists(SourceReference reference, string variableName);

   boolean Contains(string variableName);
   boolean Contains(VariableReference reference, IValidationContext context);

   VariableType GetVariableType(string variableName);
   VariableType GetVariableType(VariableReference reference, IValidationContext context);
   VariableSource? GetVariableSource(string variableName);

   VariableEntry GetVariable(string variableName);
}

File: src/Lexy.Compiler/Parser/IVariableContext.cs
using Lexy.Compiler.Language;
using Lexy.Compiler.Language.VariableTypes;
using Lexy.RunTime;

namespace Lexy.Compiler.Parser;

public interface IVariableContext
{
    void AddVariable(string variableName, VariableType type, VariableSource source);

    void RegisterVariableAndVerifyUnique(SourceReference reference, string variableName, VariableType type,
        VariableSource source);

    bool Contains(string variableName);
    bool Contains(IdentifierPath path, IValidationContext context);

    VariableType GetVariableType(string variableName);
    VariableType GetVariableType(IdentifierPath path, IValidationContext context);

    VariableEntry GetVariable(string variableName);

    VariableReference CreateVariableReference(SourceReference reference, IdentifierPath path, IValidationContext validationContext);
}